src/audio/voice/capture/voice_capture.cpp

```diff
@@ -0,0 +1,165 @@
+    capture->OnAudioCapture(buffer);
+}
+
+VoiceCapture::VoiceCapture(Context &context) noexcept :
+    m_device(context, GetDeviceConfig(), context.GetActiveCaptureID()) {}
+
+void VoiceCapture::Start() noexcept {
+    StartEncoder();
+    if (!m_encoder.Lock()->has_value()) {
+        return;
+    }
+
+    m_device.Start();
+}
+
+void VoiceCapture::Stop() noexcept {
+    m_device.Stop();
+    m_encoder.Lock()->reset();
+}
+
+void VoiceCapture::SetActive(bool active) noexcept {
+    m_active = active;
+}
+
+void VoiceCapture::SetCaptureDevice(ma_device_id &&device_id) noexcept {
+    spdlog::get("voice")->info("Setting capture device");
+
+    const auto success = m_device.ChangeDevice(std::move(device_id));
+    if (!success) {
+        spdlog::get("voice")->error("Failed to set capture device");
+    }
+}
+
+ma_device_config VoiceCapture::GetDeviceConfig() noexcept {
+    auto config = ma_device_config_init(ma_device_type_capture);
+    config.capture.format = CAPTURE_FORMAT;
+    config.sampleRate = CAPTURE_SAMPLE_RATE;
+    config.capture.channels = CAPTURE_CHANNELS;
+    config.pulse.pStreamNameCapture = "Abaddon (Capture)";
+
+    config.periodSizeInFrames = CAPTURE_FRAME_SIZE;
+    config.performanceProfile = ma_performance_profile_low_latency;
+
+    config.dataCallback = capture_callback;
+    config.pUserData = this;
+
+    return config;
+}
+
+void VoiceCapture::StartEncoder() noexcept {
+    auto settings = EncoderSettings {};
+    settings.sample_rate = CAPTURE_SAMPLE_RATE;
+    settings.channels = CAPTURE_CHANNELS;
+    settings.bitrate = 64000;
+    settings.signal_hint = SignalHint::Auto;
+    settings.application = EncodingApplication::Audio;
+
+    auto encoder = OpusEncoder::Create(settings);
+    if (!encoder) {
+        return;
+    }
+
+    m_encoder.Lock()->emplace(std::move(*encoder));
+}
+
+void VoiceCapture::OnAudioCapture(InputBuffer input) noexcept {
+    if (!m_active) {
+        return;
+    }
+
+    CaptureBuffer buffer;
+    std::copy(input.begin(), input.end(), buffer.begin());
+
+    ApplyEffects(buffer);
+    if (ApplyNoise(buffer)) {
+        EncodeAndSend(std::move(buffer));
+    }
+}
+
+void VoiceCapture::ApplyEffects(CaptureBuffer &buffer) noexcept {
+    if (m_mix_mono) {
+        AudioUtils::MixStereoToMono(buffer);
+    }
+
+    AudioUtils::ApplyGain(buffer, m_gain);
+    m_peak_meter.UpdatePeak(buffer);
+}
+
+bool VoiceCapture::ApplyNoise(CaptureBuffer &buffer) noexcept {
+    if(!m_effects.PassesVAD(buffer, m_peak_meter.GetPeak())) {
+        return false;
+    }
+
+    if (m_suppress_noise) {
+        m_effects.Denoise(buffer);
+    }
+
+    return true;
+}
+
+void VoiceCapture::EncodeAndSend(CaptureBuffer &&buffer) noexcept {
+    std::vector<uint8_t> opus;
+    opus.resize(1275);
+
+    const auto bytes = m_encoder.Lock()->value().Encode(buffer, opus, CAPTURE_FRAME_SIZE);
+    opus.resize(bytes);
+
+    if (bytes > 0) {
+        m_signal.emit(std::move(opus));
```

Comment:
emit will effectively take a const reference here so move does nothing

Reply:
i dont think sigc will do any copies if the connected slot takes it as a reference

Reply:
> i dont think sigc will do any copies

This wouldn't make sense, as the vector would get  destroyed at the end of the scope there. I tried taking `std:vector<uint8_t>&&` in the slot, but couldn't make it compile.

Reply:
u can just take a normal const reference. all slots are called when you do emit so itll be in scope

Reply:
My understanding is that sigc accepts the parameters and passes them down to all slots without waiting for callbacks to return (thus non-blocking). If that is the case, stack will get corrupted if `EncodeAndSend` scope ends before the callback returns (as it will still hold the reference to the vector). 

We need to either copy or ideally move the vector here for safe behavior.

Reply:
not sure where you got that from but sigc definitely waits. what youre describing is more like how Glib::Dispatcher works. emit also explicitly returns whatever the last connected slot returned

Reply:
Oh so it is blocking?  I couldn't find much information about it. Using `const&` should be fine then.